File: frontend/src/types/job.ts
/**
 * Shared TypeScript types for the Job domain.
 * Mirrors the backend Pydantic schemas.
 */

export type RemoteType = 'REMOTE' | 'HYBRID' | 'ONSITE'

export const REMOTE_TYPE_LABELS: Record<RemoteType, string> = {
  REMOTE: 'Remote',
  HYBRID: 'Hybrid',
  ONSITE: 'On-site',
}

export interface Job {
  id: string
  title: string
  company: string
  location: string | null
  remote_type: RemoteType | null
  salary_min: number | null
  salary_max: number | null
  description: string | null
  url: string | null
  source: string | null
  posted_date: string | null // ISO date string YYYY-MM-DD
  created_at: string
  // Computed fields — populated by the API per authenticated user
  match_score: number | null
  is_saved: boolean
}

export interface JobCreate {
  title: string
  company: string
  location?: string | null
  remote_type?: RemoteType | null
  salary_min?: number | null
  salary_max?: number | null
  description?: string | null
  url?: string | null
  source?: string | null
  posted_date?: string | null
}

export interface SavedJob {
  id: string
  user_id: string
  job_id: string
  saved_at: string
  notes: string | null
  job: Job
}

export interface SavedJobCreate {
  job_id: string
  notes?: string | null
}

export interface JobFilters {
  q?: string
  location?: string
  remote_type?: RemoteType | ''
  salary_min?: number
  salary_max?: number
  skip?: number
  limit?: number
}

/** Threshold for match score color coding per design spec */
export function getMatchScoreVariant(score: number): 'success' | 'warning' | 'default' {
  if (score >= 85) return 'success'
  if (score >= 70) return 'warning'
  return 'default'
}

/** Format salary range as a compact string, e.g. "$180–220k" */
export function formatSalaryRange(min: number | null, max: number | null): string {
  if (!min && !max) return ''
  const fmt = (n: number) => `$${Math.round(n / 1000)}k`
  if (min && max) return `${fmt(min)}–${fmt(max)}`
  if (min) return `${fmt(min)}+`
  if (max) return `Up to ${fmt(max)}`
  return ''
}

File: backend/app/schemas/job.py
"""Pydantic schemas for Job and SavedJob request/response."""

import uuid
from datetime import date, datetime
from typing import Annotated

from pydantic import BaseModel, Field

from app.models.job import RemoteType


class JobCreate(BaseModel):
    title: Annotated[str, Field(min_length=1, max_length=500)]
    company: Annotated[str, Field(min_length=1, max_length=500)]
    location: str | None = None
    remote_type: RemoteType | None = None
    salary_min: int | None = None
    salary_max: int | None = None
    description: str | None = None
    url: str | None = None
    source: str | None = "manual"
    posted_date: date | None = None


class JobUpdate(BaseModel):
    title: Annotated[str, Field(min_length=1, max_length=500)] | None = None
    company: Annotated[str, Field(min_length=1, max_length=500)] | None = None
    location: str | None = None
    remote_type: RemoteType | None = None
    salary_min: int | None = None
    salary_max: int | None = None
    description: str | None = None
    url: str | None = None
    source: str | None = None
    posted_date: date | None = None


class JobResponse(BaseModel):
    id: uuid.UUID
    title: str
    company: str
    location: str | None
    remote_type: RemoteType | None
    salary_min: int | None
    salary_max: int | None
    description: str | None
    url: str | None
    source: str | None
    posted_date: date | None
    created_at: datetime
    # Computed fields — added by the router, not stored on the model
    match_score: int | None = None
    is_saved: bool = False

    model_config = {"from_attributes": True}


class SavedJobCreate(BaseModel):
    job_id: uuid.UUID
    notes: str | None = None


class SavedJobResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    job_id: uuid.UUID
    saved_at: datetime
    notes: str | None
    job: JobResponse

    model_config = {"from_attributes": True}


class JobFilters(BaseModel):
    q: str | None = None
    location: str | None = None
    remote_type: RemoteType | None = None
    salary_min: int | None = None
    salary_max: int | None = None
    skip: int = Field(default=0, ge=0)
    limit: int = Field(default=50, ge=1, le=200)
